### PR TITLE
Rest endpoints + processing

### PR DESCRIPTION
**Pull request by CrossfireCurt**
- Introduces concept of a "Job". Jobs are what hold Pump configuration and information about Pump's progress. Users submit a Job via REST endpoints. 
- Jobs are queued in an ExecutorService. Once executed, a new Pump instance is created and run with setting defined in the Job.
- REST endpoints for interacting with Jobs; submitting, checking progress, etc. Uses Dropwizard 0.8.3 and Google Guice.
- Broke record transformation out into a Service. 
- Implemented tests for service and controller layers (more to come soon).

### CODE REVIEW

File: pump/src/main/java/com/commercehub/watershed/pump/model/Job.java [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+import org.joda.time.Period;
+import org.joda.time.format.PeriodFormatter;
+import org.joda.time.format.PeriodFormatterBuilder;
+import rx.Subscription;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class Job {
+    String jobId;
```
Comment (pmogren) [inline pump/src/main/java/com/commercehub/watershed/pump/model/Job.java]: Given that there are getters and setters I assume there is no reason for these fields to not be private.

Comment (CrossfireCurt) [inline pump/src/main/java/com/commercehub/watershed/pump/model/Job.java]: oh yeah, those are supposed to be private.

File: pump/src/main/java/com/commercehub/watershed/pump/processing/Pump.java [hunk 2]
```diff
@@ -82,43 +87,47 @@ private int countShardsInStream(String stream, KinesisProducerConfiguration kine
      * to be reported. To cancel pumping, unsubscribe.
      */
     public Observable<UserRecordResult> build() {
-        // Can't actually use rxjava-jdbc with Drill at the moment: Drill's JDBC client is broken wrt PreparedStatements (DRILL-3566)
+        // Can't actually use rxjava-jdbc with Drill at the moment: Drill'subscriber JDBC client is broken wrt PreparedStatements (DRILL-3566)
```
Comment (pmogren) [inline pump/src/main/java/com/commercehub/watershed/pump/processing/Pump.java]: Is this a find-and-replace gone wrong?

File: pump/src/main/java/com/commercehub/watershed/pump/processing/Pump.java [hunk 3]
```diff
@@ -158,6 +168,7 @@ public Record call(Record record) {
 
     void destroy() {
         if (kinesisProducer != null) {
+            kinesisProducer.flushSync();
```
Comment (pmogren) [inline pump/src/main/java/com/commercehub/watershed/pump/processing/Pump.java]: Why did you add this? As I recall, flush and destroy ended up separate for a reason.

File: pump/src/main/java/com/commercehub/watershed/pump/resources/PumpResource.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.io.IOException;
+import java.util.Collection;
+
+
+@Path("/pump")
+@Produces(MediaType.APPLICATION_JSON)
+public class PumpResource {
+final Logger log = LoggerFactory.getLogger(PumpResource.class);
```
Comment (pmogren) [inline pump/src/main/java/com/commercehub/watershed/pump/resources/PumpResource.java]: Shouldn't these fields be private?

File: pump/src/main/java/com/commercehub/watershed/pump/service/TransformerServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayOutputStream;
+
+public class TransformerServiceImpl implements TransformerService {
+    private static final Logger log = LoggerFactory.getLogger(TransformerServiceImpl.class);
+
+    @Inject
+    ObjectMapper objectMapper;
```
Comment (pmogren) [inline pump/src/main/java/com/commercehub/watershed/pump/service/TransformerServiceImpl.java]: private?

File: pump/src/main/java/com/commercehub/watershed/pump/application/PumpGuiceModule.java [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    private Pump pumpProvider(Database database, KinesisProducerConfiguration kinesisProducerConfiguration){
+        return new Pump(database, kinesisProducerConfiguration);
+    }
+
+
+    @Provides
+    @Singleton
+    private KinesisProducerConfiguration configureKinesis() {
+        KinesisProducerConfiguration kinesisConfig = new KinesisProducerConfiguration();
+
```
Comment (pmogren) [inline pump/src/main/java/com/commercehub/watershed/pump/application/PumpGuiceModule.java]: All of these options should come from config file

File: pump/src/main/java/com/commercehub/watershed/pump/processing/JobRunnable.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+import rx.Observable;
+import rx.Subscriber;
+import rx.Subscription;
+
+import java.text.NumberFormat;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class JobRunnable implements Runnable {
+    private static final Logger log = LoggerFactory.getLogger(JobRunnable.class);
+    private static final int RECORDS_PER_CHUNK = 1000;
```
Comment (pmogren) [inline pump/src/main/java/com/commercehub/watershed/pump/processing/JobRunnable.java]: This should probably be configurable too

File: pump/src/main/java/com/commercehub/watershed/pump/processing/Pump.java [hunk 8]
```diff
@@ -82,43 +87,47 @@ private int countShardsInStream(String stream, KinesisProducerConfiguration kine
...
+            public void call(final Subscriber<? super Record> subscriber) {
+                subscriber.onStart();
+                final Connection connection = getConnection(subscriber);
+
+                final ResultSet resultSet = executeQuery(subscriber, connection);
+
+                JdbcRecordProducer jdbcRecordProducer = new JdbcRecordProducer(subscriber, resultSet, connection, pumpSettings);
+                subscriber.setProducer(jdbcRecordProducer);
+
                 jdbcRecordProducer.request(shardCount * MAX_RECORDS_PER_SHARD_PER_SECOND * kinesisConfig.getRateLimit() / 200);
```
Comment (pmogren) [inline pump/src/main/java/com/commercehub/watershed/pump/processing/Pump.java]: MAX_RECORDS_PER_SHARD_PER_SECOND should probably be configurable too, so we can compensate for aggregation

File: pump/src/main/java/com/commercehub/watershed/pump/application/PumpGuiceModule.java [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+        kinesisConfig.setAggregationEnabled(false); //TODO enable KPL aggregation after Filter's KCL is upgraded; it should be a big help.
```
Comment (pmogren) [inline pump/src/main/java/com/commercehub/watershed/pump/application/PumpGuiceModule.java]: remove mention of Filter

File: pump/src/main/java/com/commercehub/watershed/pump/resources/PumpResource.java [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+    @Inject
+    JobService jobService;
+
+    @Inject
+    public PumpResource() {
+        log.info("Creating a new PumpResource.");
+    }
+
+    @Path("/status/{job_id}")
```
Comment (pmogren) [inline pump/src/main/java/com/commercehub/watershed/pump/resources/PumpResource.java]: Why not use /jobs in place of /status and /queue? Seems more RESTful.

File: pump/src/main/java/com/commercehub/watershed/pump/service/JobService.java [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+
+import com.commercehub.watershed.pump.model.Job;
+import com.commercehub.watershed.pump.model.JobPreview;
+import com.commercehub.watershed.pump.model.PreviewSettings;
+import com.commercehub.watershed.pump.model.PumpSettings;
+
+import java.util.Collection;
+
+public interface JobService {
+    Job queueJob(PumpSettings pumpSettings);
```
Comment (pmogren) [inline pump/src/main/java/com/commercehub/watershed/pump/service/JobService.java]: Please rename queueJob to enqueueJob

**Comment by CrossfireCurt** [comment]
@pmogren all set!
